resources/shaders/basic.frag

```diff
@@ -1,27 +1,28 @@
-#version 150
+#version 330
 precision mediump float;
 
 in vec2 uv;
+flat in vec2 imageSize;
+flat in vec4 hoverColor;
+flat in vec3 atlasOffset;
+
 out vec4 frag_colour;
-uniform sampler2D tex;
-uniform float h_color_r;
-uniform float h_color_g;
-uniform float h_color_b;
-uniform float h_color_a;
-uniform float imgW;
-uniform float imgH;
+uniform sampler2D tex[8]; // NOTE: The size of this array has a large impact on performance...
```

Comment:
This array size has the main impact on performance.
With a `GL_MAX_TEXTURE_SIZE` of 16384 this only needs to be 2, 8192 is still common though so went with 8.
Using a `GL_TEXTURE_2D_ARRAY` avoids the need for this array (`uniform sampler2DArray tex;`)